src/Foundatio.Redis/Cache/RedisCacheClient.cs

```diff
@@ -80,52 +84,50 @@ public async Task<int> RemoveAllAsync(IEnumerable<string> keys = null)
 
                 try
                 {
+                    long dbSize = await server.DatabaseSizeAsync().AnyContext();
                     await server.FlushDatabaseAsync().AnyContext();
+                    deleted += dbSize;
                     continue;
                 }
                 catch (Exception ex)
                 {
-                    if (_logger.IsEnabled(LogLevel.Error))
-                        _logger.LogError(ex, "Unable to flush database");
+                    _logger.LogError(ex, "Unable to flush database: {Message}", ex.Message);
                 }
 
                 try
                 {
+                    // NOTE: We need to use a HashSet to avoid duplicate counts due to SCAN is non-deterministic.
+                    // A Performance win could be had if we are sure dbSize didn't fail and we know nothing was changing
+                    // keys while we were deleting.
+                    var seen = new HashSet<RedisKey>();
                     await foreach (var key in server.KeysAsync().ConfigureAwait(false))
-                        await Database.KeyDeleteAsync(key).AnyContext();
+                        seen.Add(key);
+
+                    foreach (var batch in seen.Batch(batchSize))
+                        deleted += await Database.KeyDeleteAsync(batch.ToArray()).AnyContext();
                 }
                 catch (Exception ex)
                 {
-                    if (_logger.IsEnabled(LogLevel.Error))
-                        _logger.LogError(ex, "Unable to delete all");
+                    _logger.LogError(ex, "Error deleting all keys: {Message}", ex.Message);
                 }
             }
         }
         else
         {
-            var redisKeys = keys.Where(k => !String.IsNullOrEmpty(k)).Select(k => (RedisKey)k).ToArray();
-            if (redisKeys.Length <= 0)
-                return 0;
-
-            int count = 0;
-            foreach (var key in redisKeys)
+            foreach (var redisKeys in keys.Where(k => !String.IsNullOrEmpty(k)).Select(k => (RedisKey)k).Batch(batchSize))
```

Comment:
Batch size matches the size that SE client uses internally for getting batches of keys. So I just reused it, higher count like 1k went faster but then it's basically getting batches and sending batches with better alignment.